src/shared/cloudformation/activation.ts

```diff
@@ -0,0 +1,66 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+import { associateFileSystemWatcherWithListener } from '../utilities/fileSystemWatcher'
+import {
+    CloudFormationTemplateRegistry,
+    DefaultCloudFormationTemplateRegistry,
+    DefaultCloudFormationTemplateRegistryListener,
+    setTemplateRegistry
+} from './templateRegistry'
+
+// this is not a relative pattern and thus will search against all workspace folders.
+// also note: GlobPatterns do not respect `\`. Therefore, do not use `path.join` to create the path.
+const TEMPLATE_FILE_GLOB_PATTERN = '**/template.{yaml,yml}'
+
+/**
+ * Creates a CloudFormationTemplateRegistry which retains the state of CloudFormation templates in a workspace.
+ * This also assigns a FileSystemWatcher which will update the registry on any change to tracked templates.
+ *
+ * @param extensionContext VS Code extension context
+ * @param globPattern Glob pattern for files to track. Default: TEMPLATE_FILE_GLOB_PATTERN
+ */
+export async function activate(
+    extensionContext: vscode.ExtensionContext,
+    globPattern: string = TEMPLATE_FILE_GLOB_PATTERN
+): Promise<void> {
+    const registry = new DefaultCloudFormationTemplateRegistry()
+    const listener = new DefaultCloudFormationTemplateRegistryListener(registry)
+    const watcher = vscode.workspace.createFileSystemWatcher(globPattern)
+    associateFileSystemWatcherWithListener(watcher, listener, globPattern)
+    const templatePaths = await vscode.workspace.findFiles(globPattern)
+    await populateRegistry(registry, templatePaths)
+    setTemplateRegistry(registry)
+
+    extensionContext.subscriptions.push(listener)
+    extensionContext.subscriptions.push(watcher)
+}
+
+/**
+ * Initial registry population is an asynchronous task and cannot be initiated in the constructor
+ * Call this function immediately after constructing the registry class or else you'll have an empty registry
+ * Handle as many template files as possible, as quickly as possible by queuing up all promises immediately
+ *
+ * @param registry CloudFormationTemplateRegistry to initially populate
+ * @param templatePaths Paths to enter into the registry
+ */
+export async function populateRegistry(registry: CloudFormationTemplateRegistry, templatePaths: vscode.Uri[]): Promise<void> {
+    const templateParsingPromises: Promise<void>[] = []
+    for (const templatePath of templatePaths) {
+        templateParsingPromises.push(
+            new Promise(async (resolve) => {
+                try {
+                    await registry.addTemplateToTemplateData(templatePath)
```

Comment:
why is this wrapped in a promise? Is `addTemplateToTemplateData()` doing expensive out-of-process work?

Reply:
Yes, this is opening and parsing a template file. This is wrapped in a try/catch so `Promise.all` doesn't break if the file is not parseable YAML.

This is a replacement for [`promise.allSettled()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) which is not present in our Node version.

Reply:
> Yes, this is opening and parsing a template file

What is doing the parsing? If the parsing is done in the node.js process , nothing is gained by promisfying and parallelizing it. 

And even if the parsing were done out-of-process, it's still not worth the code complication: parsing a file takes a few milliseconds.